### PR TITLE
Input transformation tool 

### PR DESCRIPTION
**Pull request by dschwen**
I'd like to get some input here. In particular, where to put the thing and how to integrate it into the build process. Right now it is

```
cd $MOOSE_DIR/framework/contrib/hit
make rewrite
```

I also have a few hit issues that become apparent when doing this type of stuff. 
* Removal of `../` and `./` by `render` inflates the diff
* Root-level Blank lines are not preserved (yet at section level they are)

Refs #10368

### CODE REVIEW

**Comment by rwcarlsen** [comment]
@dschwen - can you see if #10371 resolves the minor rendering issues you brought up?

**Comment by dschwen** [comment]
Check b3e19d5 this is a bit messed up. The `[../]` is used to close top level sections now. The subsections are still closed by `[]`. Blank lines are still missing.

**Comment by rwcarlsen** [comment]
#10371 is in and should fix the rendering issues.

**Comment by dschwen** [comment]
@rwcarlsen, yes, this is great. The diff between input and output is now minimal:

```
56,61d55
<   [./left]
<     type = FunctionDirichletBC
<     variable = disp_x
<     boundary = 3
<     function = 0.02*t
<   [../]
70,71c64,71
<   [./back]
<     type = FunctionDirichletBC
---
>   [left]
>     type = FunctionPresetBC
>     variable = disp_x
>     function = 0.02*t
>     boundary = 3
>   []
>   [back]
>     type = FunctionPresetBC
73d72
<     boundary = 1
75c74,75
<   [../]
---
>     boundary = 1
>   []
```

It is limited to what the transformation rules state 👍 .

File: framework/contrib/hit/rewrite.cc [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+      if (!matchSection(match, input, delete_list, matched_params))
+        break;
+
+      // delete what can be deleted
+      for (auto * del : delete_list)
+      {
+        if (del->type() == NodeType::Field)
+          delete del;
+        else if (del->type() == NodeType::Section && del->children(NodeType::All).empty())
+          delete del;
```
Comment (permcody) [inline framework/contrib/hit/rewrite.cc]: 😭

File: framework/contrib/hit/rewrite.cc [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+      // insert replacement
+      hit::merge(replacement, input);
+    }
+  }
+
+  std::cout << input->render() << '\n';
+
+  // clean up
+  delete input;
+  delete rules_root;
```
Comment (permcody) [inline framework/contrib/hit/rewrite.cc]: 😭

File: framework/contrib/hit/rewrite.cc [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+  // load first rule file
+  auto * rules_root = loadAndParse(argv[2]);
+
+  // load additional rule files
+  for (unsigned int i = 3; i < argc; ++i)
+  {
+    auto * more_rules = loadAndParse(argv[i]);
+    merge(more_rules, rules_root);
+    delete more_rules;
```
Comment (permcody) [inline framework/contrib/hit/rewrite.cc]: 😭

**Review by rwcarlsen — APPROVED** [review]
I'm okay with this just living in the contrib/hit dir with a ``rewrite`` target.  Seems reasonable for the use cases of the tool - and it is somewhat generically applicable to hit in general - not just necessarily moose syntax.

**Pull request merged**